drivers/usb-common.c

```diff
@@ -410,3 +410,80 @@ void warn_if_bad_usb_port_filename(const char *fn) {
 		__func__, fn);
 	return;
 }
+
+/* API neutral, handles retries */
+static int nut_usb_get_string_descriptor(
+	usb_dev_handle *udev,
+	int StringIdx,
+	int langid,
+	char *buf,
+	size_t buflen)
+{
+	int ret = -1;
+	int retries = 3;
+
+	while (retries--) {
+		ret = usb_get_string(udev, (usb_ctrl_strindex)StringIdx, langid, (usb_ctrl_charbuf)buf, buflen);
+		if (ret >= 0) {
+			break;
+		} else if (retries) {
+			upsdebugx(1, "%s: string descriptor %d request failed, retrying...", __func__, StringIdx);
+			usleep(50);
+		}
+	}
+	return ret;
+}
+
+/* API neutral, assumes en_US if langid descriptor is broken */
+int nut_usb_get_string(
+	usb_dev_handle *udev,
+	int StringIdx,
+	char *buf,
+	size_t buflen)
+{
+	int ret;
+	char buffer[255];
+	int langid;
+	int len;
+	int i;
+
+	if (!udev || StringIdx < 1 || StringIdx > 255) {
+		return -1;
+	}
+
+	/* request langid descriptor */
+	ret = nut_usb_get_string_descriptor(udev, 0, 0, buffer, 4);
+	if (ret < 0)
+		return ret;
+
+	if (ret == 4 && buffer[0] >= 4 && buffer[1] == USB_DT_STRING) {
+		langid = buffer[2] | (buffer[3] << 8);
+	} else {
+		upsdebugx(1, "%s: Broken language identifier, assuming en_US", __func__);
+		langid = 0x0409;
```

Comment:
Yes, I think ideally the fallback value should be configurable, in case some (broken) device shows up and requires another fallback value. But this can safely wait until such a device is proven to exist maybe? I think there will be (more) non-en_US devices in the future, but at the same time their manufacturers will probably make sure their langid's are correct in this case.

BTW, with regards to possibly generalizing e.g. langid_fix. I don't know how nut works with multiple configs and multiple devices. Will each device have its own process, so globals are fine, and will get_var only return the values relevant for this instance?

And no, libusb has no langid mappings or helpers. It doesn't touch langid other than in the _simple/_ascii helpers (apart from possibly some assumptions in the little used Windows HID backend).

Reply:
Happy to see that things are good.
Congrats for the team work.
And thanks for the credits @tormodvolden